src/plugins/intel_cpu/src/mkldnn_node.cpp

```diff
@@ -1418,43 +1419,68 @@ bool MKLDNNNode::needShapeInfer() const {
 }
 
 std::vector<VectorDims> MKLDNNNode::shapeInfer() const {
-    std::vector<Shape> shapes;
-    for (size_t i = 0; i < inputShapes.size(); i++) {
-        shapes.push_back(getParentEdgesAtPort(i)[0]->getMemory().getDesc().getShape());
+    return shapeInferGeneric();
+}
+
+std::vector<VectorDims> MKLDNNNode::shapeInferGeneric(const std::vector<ov::StaticShape>& input_shapes,
+                                                      uint32_t input_value_port_mask) const {
+    // collect input values
+    std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>> input_values;
+    if (input_value_port_mask) {
+        for (size_t port = 0; port < inputIsScalar.size(); port++) {
```

Comment:
What about the following. `IShapeInfer` is abstract, and the user passes some `input_shapes` considering that the number of these shapes and their ranks are what is expected by the underlying operation. What if we could ask `IShapeInfer` about what number of inputs and what ranks is expected? So we were able to feed it with valid data. What do you think, does it make sense to add `getInputRanks()` method to `IShapeInfer` so we can check the inputs number by the size of the return vector and their ranks (the members of the return vector), the latter allows us to handle scalars?

Reply:
So you can see here we are facing 2 problems:

1. we don't know how many inputs the shape_inference API is expecting since its based on ngraph OP, but MKLDNNNode with fused operation introduces new inputs so we may provide too many input shapes than expected, and that breaks some assert inside shape_infer API.

2. the scalar shape checks are too strict in shape infer.

in order to fix it, we can:
OptionA: satisfy shape inference API's expectation as much as possible, that's what you are saying by adding `vector<size_t> getInputRanks()`,

OptionB: we can loose the assertion inside shape infer API so additional input shapes are allowed and length-1 1D vector is treated as equivalent to scalar.

I personally prefer to OptionB since it makes the API more flexible and robust, also performance is better since less negotiation is required,  otherwise, if we go OptionA, both code complexity and run-time cost may be increased.

What do you think?

Reply:
Also why `MKLDNNNode::MKLDNNNode()` use length-1 1D to replace scalar ?
```
bool isScalar = shape.rank().get_length() == 0;
inputShapes.emplace_back(isScalar ? ngraph::PartialShape{1} : shape);
```

Reply:
I'm asking because OptionB may face difficulty because by ngraph OP definition. some input shape should be scalar {} instead of {1}

Reply:
Quick answer: option B looks better than A, but we need to estimate consequences of loosing assertions.
Long answer:
From the performance standpoint, adding `getInputRanks()` will not give us extra overhead in comparison with the current implementation since we just change the `inputIsScalar` vector with the result of `getInputRanks()` and the resulting vector of `getInputRanks()` can be cached inside `IShapeInfer` since it does not change.
On the other hand, even if we add `getInputRanks()` it will not force the user to check the `input_shape`, it only provides an extra info that may or may not be considered, that means it does not reduce the error prone rate, its only purpose that I see at the moment. The whole problem is if we somehow feed the `infer` call with a wrong order of inputs, I mean wrong order of inputs with same ranks, we will silently fail with a wrong result shape and it will be hard to detect.  Of course `getInputRanks()` does not solve the problem. And it looks like we can not solve it in the current paradigm moving all the responsibility of input data validation to the user.
Regarding the code complexity, it looks temptive to pass just all the inputs and let the `IShapeInfer` implementation decide what specific shapes it needs but passing extra shapes means extra conversions from `Shape` to `StaticShape`. Regrading preventing errors I think it does not change anything. As for loosing requirements to a scalar description, I mean accepting both zero rank tensors and first rank tensors of shape {1} as a scalar, it does not look dangerous, but I think we need to carefully estimate possible consequences in terms of error-prone probability.

Reply:
> Also why `MKLDNNNode::MKLDNNNode()` use length-1 1D to replace scalar ?
> 
> ```
> bool isScalar = shape.rank().get_length() == 0;
> inputShapes.emplace_back(isScalar ? ngraph::PartialShape{1} : shape);
> ```

That is by historical reasons. To solve the problem we have to update the whole plugin at once.

Reply:
> I'm asking because OptionB may face difficulty because by ngraph OP definition. some input shape should be scalar {} instead of {1}

If we plan to move this responsibility to the `IShapeInfer` we will have to replace {1} tensors with scalars somewhere inside the  `IShapeInfer` implementations. Alternative is the plan A.

Reply:
OK, that's good idea, I didn't realize that we can cache the input ranks in IShapeInfer, that dispels my concern on performance, I will add that API